stabilize combining +bundle and +whole-archive link modifiers

Be-ing opened the pull request:
Per discussion on https://github.com/rust-lang/rust/issues/108081 combining +bundle and +whole-archive already works and can be stabilized independently of other aspects of the packed_bundled_libs feature. There is no risk of regression because this was not previously allowed.

r? @petrochenkov

Be-ing committed:
stabilize combining +bundle and +whole-archive link modifiers

Currently, combining +bundle and +whole-archive works only with
 #![feature(packed_bundled_libs)]
This crate feature is independent of the -Zpacked-bundled-libs
command line option.

This commit stabilizes the #![feature(packed_bundled_libs)] crate
feature and implicitly enables it only when the +bundle and
+whole-archive link modifiers are combined. This allows rlib
crates to use the +whole-archive link modifier with native
libraries and have all symbols included in the linked library
to be included in downstream staticlib crates that use the rlib as
a dependency. Other cases requiring the packed_bundled_libs
behavior still require the -Zpacked-bundled-libs command line
option, which can be stabilized independently in the future.

Per discussion on https://github.com/rust-lang/rust/issues/108081
there is no risk of regression stabilizing the crate feature in
this way because the combination of +bundle,+whole-archive link
modifiers was previously not allowed.

rust-log-analyzer commented:
The job **`mingw-check-tidy`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/5447943823/jobs/9910547870) [(plain)](https://github.com/rust-lang/rust/commit/694a0489e6d68267e8af2b36f4a073591618b2c9/checks/14748216393/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Prepare all required actions
Getting action download info
Download action repository 'actions/checkout@v3' (SHA:c85c95e3d7251135ab7dc9ce3241c5835cc595a9)
Download action repository 'actions/upload-artifact@v3' (SHA:0b7f8abb1508181956e8e162db84b466c27e18ce)
Complete job name: PR - mingw-check-tidy
git config --global core.autocrlf false
shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
env:
  CI_JOB_NAME: mingw-check-tidy
---
Building wheels for collected packages: reuse
  Building wheel for reuse (pyproject.toml): started
  Building wheel for reuse (pyproject.toml): finished with status 'done'
  Created wheel for reuse: filename=reuse-1.1.0-cp310-cp310-manylinux_2_35_x86_64.whl size=180116 sha256=351235b2326fb4db7a18e257e13ce7896c5f77339521e2c2612e71e154800a19
  Stored in directory: /tmp/pip-ephem-wheel-cache-d21rrv04/wheels/c2/3c/b9/1120c2ab4bd82694f7e6f0537dc5b9a085c13e2c69a8d0c76d
Installing collected packages: boolean-py, binaryornot, setuptools, reuse, python-debian, markupsafe, license-expression, jinja2, chardet
  Attempting uninstall: setuptools
    Found existing installation: setuptools 59.6.0
    Not uninstalling setuptools at /usr/lib/python3/dist-packages, outside environment /usr
---
Successfully tagged rust-ci:latest
Built container sha256:730ae5b50a11297c0e0316a6abcd958ad310469e6a77efd65400c02f9a5730fd
Uploading finished image to https://ci-caches.rust-lang.org/docker/39289e18f1ac1b1a573d1a3c658528cab0cbfec4029e2def8ed615ecd4b1295e23213ab5b134f4a50a4bd45af0a8a4266b7571ba256dc217696d0f040efcd003

<botocore.awsrequest.AWSRequest object at 0x7fbc07d40350>
gzip: stdout: Broken pipe
xargs: docker: terminated by signal 13
[CI_JOB_NAME=mingw-check-tidy]
[CI_JOB_NAME=mingw-check-tidy]
---
##[endgroup]
fmt check
tidy check
tidy: Skipping binary file check, read-only filesystem
Expected a gate test for the feature 'packed_bundled_libs'.
Hint: create a failing test file named 'tests/ui/feature-gates/feature-gate-packed-bundled-libs.rs',
      with its failures due to missing usage of `#![feature(packed_bundled_libs)]`.
Hint: If you already have such a test and don't want to rename it,
      you can also add a // gate-test-packed_bundled_libs line to the test file.
tidy error: Found 1 features without a gate test.
some tidy checks failed
```

</details>

petrochenkov commented on compiler/rustc_codegen_ssa/src/back/link.rs:
```diff
@@ -365,15 +365,9 @@ fn link_rlib<'a>(
         };
-        if whole_archive == Some(true)
-            && flavor == RlibFlavor::Normal
-            && !codegen_results.crate_info.feature_packed_bundled_libs
```
The `feature_packed_bundled_libs` field can be removed too, it was its only use.

petrochenkov commented:
The `packed_bundled_libs` feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).

Be-ing commented:
> The packed_bundled_libs feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).

What about the [other use](https://github.com/rust-lang/rust/blob/3ae0ef79fe4485ad5f0bfc4cd1f27a1b3fd94f60/compiler/rustc_metadata/src/native_libs.rs#L48-L67) of the feature?

petrochenkov commented:
> > The packed_bundled_libs feature should also be stabilized (moved from compiler\rustc_feature\src\active.rs to compiler\rustc_feature\src\accepted.rs).
> 
> What about the [other use](https://github.com/rust-lang/rust/blob/3ae0ef79fe4485ad5f0bfc4cd1f27a1b3fd94f60/compiler/rustc_metadata/src/native_libs.rs#L48-L67) of the feature?

That's a command line option, not feature.

Be-ing commented:
Ah, I didn't realize those were turned on in different ways.

Be-ing commented:
Sorry for letting this sit for a while. @petrochenkov  I have rebased it and addressed review comments.

### Brief summary
Currently, combining +bundle and +whole-archive works only with `#![feature(packed_bundled_libs)]`. This crate feature is independent of the `-Zpacked-bundled-libs` command line option.

This commit stabilizes the `#![feature(packed_bundled_libs)]` crate feature and implicitly enables it only when the +bundle and +whole-archive link modifiers are combined. This allows rlib crates to use the +whole-archive link modifier with native
libraries and have all symbols included in the linked library to be included in downstream staticlib crates that use the rlib as a dependency. Other cases requiring the packed_bundled_libs behavior still require the -Zpacked-bundled-libs command line option, which can be stabilized independently in the future.

Per discussion on https://github.com/rust-lang/rust/issues/108081 there is no risk of regression stabilizing the crate feature in this way because the combination of +bundle,+whole-archive link modifiers was previously not allowed.

### Documentation
I don't think anything needs to be updated for this? From users' perspective it's merely allowing a combination of existing link modifiers that previously failed with an error.

### Tests
[tests/run-make/rlib-format-packed-bundled-libs-3/rust_dep.rs](https://github.com/rust-lang/rust/tree/72e29da3eccd3e4c1fb2c581fa33216db50fcc93/tests/run-make/rlib-format-packed-bundled-libs-3)

Real world example of Rust code requiring this feature: https://github.com/KDAB/cxx-qt/pull/598

### Unresolved questions
The `-Zpacked-bundled-libs` command line option is still unstable.

petrochenkov commented:
@rfcbot fcp merge
This PR is a promise that the combination `+bundle,+whole-archive` in rlibs will be supported in the future in some way (the specific way in which it's supported stays an implementation detail).

apiraino commented:
Discussed ion T-compiler meeting [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.5D.202023-08-17/near/385686940), acknowledged that it will need sign-off from T-compiler members.

@rustbot label -I-compiler-nominated

petrochenkov commented:
@bors r+

bors commented:
:pushpin: Commit 72e29da3eccd3e4c1fb2c581fa33216db50fcc93 has been approved by `petrochenkov`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=petrochenkov 72e29da3eccd3e4c1fb2c581fa33216db50fcc93 -->
<!-- homu: {"type":"Approved","sha":"72e29da3eccd3e4c1fb2c581fa33216db50fcc93","approver":"petrochenkov","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 72e29da3eccd3e4c1fb2c581fa33216db50fcc93 with merge 56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7...
<!-- homu: {"type":"BuildStarted","head_sha":"72e29da3eccd3e4c1fb2c581fa33216db50fcc93","merge_sha":"56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/6353968026/job/17262917471)
Approved by: petrochenkov
Pushing 56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"petrochenkov","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/6353968026/job/17262917471"},"merge_sha":"56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7"} -->

rust-timer commented:
Finished benchmarking commit (56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7): [comparison URL](https://perf.rust-lang.org/compare.html?start=a6dfd89fa76e5dda36d07463d0e54268d6240b49&end=56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=a6dfd89fa76e5dda36d07463d0e54268d6240b49&end=56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=a6dfd89fa76e5dda36d07463d0e54268d6240b49&end=56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -2.8% | [-2.8%, -2.8%] | 1     |
| Improvements ✅ <br /> (secondary) | -3.1% | [-4.4%, -1.8%] | 2     |
| All ❌✅ (primary)                 | -2.8% | [-2.8%, -2.8%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=a6dfd89fa76e5dda36d07463d0e54268d6240b49&end=56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7&stat=cycles:u)
This benchmark run did not return any relevant results for this metric.

### [Binary size](https://perf.rust-lang.org/compare.html?start=a6dfd89fa76e5dda36d07463d0e54268d6240b49&end=56ada88e7e1f00e6c6e8a0957de295bc0e2d64f7&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 631.332s -> 630.734s (-0.09%)
**Artifact size**: 317.36 MiB -> 317.40 MiB (0.01%)
